### PR TITLE
Tpetra: remove unnecessary copy to host in TAFC

### PR DESCRIPTION
**Pull request by lucbv**
@trilinos/tpetra 

## Motivation
This provides further performance improvements on device for TAFC and simplifies the code.
It's also a step toward full device execution of TAFC where it constructs the CrsMatrix.
That's an important update as the data associated with the CrsMatrix is potentially quite large and moving it back and forth on host and device is really a performance issue.

## Related Issues

* Closes 
* Blocks 
* Is blocked by 
* Follows #11751 
* Precedes 
* Related to 
* Part of #11689 
* Composed of 

## Stakeholder Feedback
N/A

## Testing
Tested on weaver with complex enabled.

### CODE REVIEW

File: packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp [hunk 1]
```diff
@@ -8588,6 +8588,24 @@ CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
...
+    // using said views as input. This should reduce the
+    // amount of data transfers required between host and
+    // device and thus improve performance.
+    typename row_ptrs_host_view_type::non_const_type            CSR_rowptr_host(CSR_rowptr.getRawPtr(), CSR_rowptr.size());
+    typename nonconst_local_inds_host_view_type::non_const_type CSR_colind_LID_host(CSR_colind_LID.getRawPtr(), CSR_colind_LID.size());
+    typename nonconst_values_host_view_type::non_const_type     CSR_vals_host(reinterpret_cast<impl_scalar_type*>(CSR_vals.getRawPtr()), CSR_vals.size());
+
+    typename row_ptrs_device_view_type::non_const_type   CSR_rowptr_dev("rowmap", CSR_rowptr.size());
+    typename local_inds_device_view_type::non_const_type CSR_colind_LID_dev("colind", CSR_colind_LID.size());
+    typename values_device_view_type::non_const_type     CSR_vals_dev("values", CSR_vals.size());
```
Comment (brian-kelley) [inline packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp:8603]:
```suggestion
    typename row_ptrs_device_view_type::non_const_type   CSR_rowptr_dev(Kokkos::ViewAllocateWithoutInitializing("rowmap"), CSR_rowptr.size());
    typename local_inds_device_view_type::non_const_type CSR_colind_LID_dev(Kokkos::ViewAllocateWithoutInitializing("colind"), CSR_colind_LID.size());
    typename values_device_view_type::non_const_type     CSR_vals_dev(Kokkos::ViewAllocateWithoutInitializing("values"), CSR_vals.size());
```

**Review by brian-kelley — APPROVED** [review]
Nice, now I see what you had planned in the last PR :)
Only suggestion is you could skip initializing the device views.

**Comment by lucbv** [comment]
Thanks @brian-kelley your change is committed, it makes a lot of sense.
Eventually, @jhux2 is working on similar changes that will allow us to remove completely the reliance on the ArrayRCP created on lines 8466 to 8469.
We are hoping to see a decent amount of performance improvement.

**Review by brian-kelley — DISMISSED** [review]

**Review by jhux2 — APPROVED** [review]
Lgtm, thanks!

**Review by jhux2 — APPROVED** [review]
lgtm

**Pull request merged**